src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/IProjectServiceExtensions.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using Microsoft.VisualStudio.Shell.Interop;
+using static Microsoft.VisualStudio.VSConstants;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    /// <summary>
+    /// Extension methods for <see cref="IProjectService"/>.
+    /// </summary>
+    internal static class IProjectServiceExtensions
+    {
+        /// <summary>
+        /// Obtains the <see cref="UnconfiguredProject"/> for <paramref name="vsHierarchy"/>,
+        /// optionally testing whether it has a capability match to <paramref name="appliesToExpression"/>.
+        /// </summary>
+        /// <param name="projectService">The project service to invoke this method on.</param>
+        /// <param name="vsHierarchy">The VS hierarchy object, which must represent a CPS project.</param>
+        /// <param name="appliesToExpression">An optional project capability match expression to filter out non-matching projects.</param>
+        /// <returns>
+        /// The CPS <see cref="UnconfiguredProject"/> for the given hierarchy node, matching any (optional)
+        /// capability expression, or <see langword="null"/> if no suitable project exists.
+        /// </returns>
+        public static UnconfiguredProject? GetUnconfiguredProject(this IProjectService projectService, IVsHierarchy vsHierarchy, string? appliesToExpression = null)
+        {
+            // We need IProjectService2.GetLoadedProject.
+            if (projectService is IProjectService2 projectService2)
+            {
+                vsHierarchy.GetCanonicalName((uint)VSITEMID.Root, out string? projectFilePath);
+
+                // Ensure we have a path.
+                if (!Strings.IsNullOrEmpty(projectFilePath))
+                {
+                    UnconfiguredProject? unconfiguredProject = projectService2.GetLoadedProject(projectFilePath);
```

Comment:
This is the critical part, right? This will only return the `UnconfiguredProject` if it is _already_ loaded?

Reply:
Yes. And importantly it happens without invoking any code on the hierarchy other than querying for the canonical name. This makes it very unlikely we inadvertently trigger any code paths that force loads.